### DISCUSSION TITLE
How to use Events in a reusable way?

### DISCUSSION THREAD

**Discussion by Barbacamanitu** [vizia/vizia#410] [Questions]
I'm following the getting started guide, and I'm on the "Making the Counter Reusable" page at: https://book.vizia.dev/quickstart/components.html

It shows how to build the ui elements into their own view and how to pass in a lens. However, it seems like the events that the buttons emit must be hardcoded at the application level so that the app can respond to them. Is it possible to define the event enum inside the custom view and still respond to this event type from an app, potentially in another crate.

**Comment by geom3trik** [comment]
Yes there's a couple of possibilities here. Firstly, I should note that in this case the counter can't really be a self-contained component, because the data which it binds to is external to the component and has to be mutated by the model. I might change the language in the guide to reflect this because I know that in other frameworks a component is completely self-contained and is also responsible for mutating the state.

Anyway, on to the solution. There's two approaches which work equally well here.

1. Define an event enum with the component (in the same module) and have the component emit this. Then respond to these events in the app data model. (E.g. `CounterEvent`)
2. Add callbacks to the counter component which are triggered when the counter buttons are pressed. This allows the user of the component to emit their own events and the counter doesn't need to know about any events at all. E.g.

```rust
Counter::new(cx, AppData::count)
    .on_increment(|cx| cx.emit(AppEvent::Increment));
```

In both cases it's still up to the model to respond to events and actually mutate the data. Which you choose depends on how you want to modularise your code. In the first the app data needs to know about the counter events. I personally prefer the latter because it means I can have events which relate specifically to the model (e.g. `AppData` and `AppEvent`). In either case the component can then be completely separate, e.g. from a separate crate.

I'll update the guide with these suggestions.
